paml_convert/markdown/markdown_specialization.py

```diff
@@ -0,0 +1,184 @@
+            value = i.value.value
+            if parameter.direction == uml.PARAMETER_OUT:
+                output_parameters.append(f"`{parameter.name}` from `{value}`")
+        output_parameters = ", ".join(output_parameters)
+        return f"Report values for {output_parameters}."
+
+    def define_container(self, record: paml.ActivityNodeExecution):
+        results = {}
+        call = record.call.lookup()
+        parameter_value_map = call.parameter_value_map()
+
+        spec = parameter_value_map["specification"]['value']
+        samples_var = parameter_value_map["samples"]['value']
+
+        ## Define a container
+
+        l.debug(f"define_container:")
+        l.debug(f" specification: {spec}")
+        l.debug(f" samples: {samples_var}")
+
+        possible_container_types = matching_containers(spec)
+        containers_str = ",".join([f"\n\t[{c.split('#')[1]}]({c})" for c in possible_container_types])
+
+        self.markdown_steps += [f"Provision a container named `{samples_var.name}` such as: {containers_str}."]
+
+        return results
+
+    # def provision_container(self, wells: WellGroup, amounts = None, volumes = None, informatics = None) -> Provision:
+    def provision_container(self, record: paml.ActivityNodeExecution):
+        results = {}
+        call = record.call.lookup()
+        parameter_value_map = call.parameter_value_map()
+
+        destination = parameter_value_map["destination"]["value"]
+        #dest_wells = self.var_to_entity[destination]
+        value = parameter_value_map["amount"]["value"].value
+        units = parameter_value_map["amount"]["value"].unit
+        units = tyto.OM.get_term_by_uri(units)
+        resource = parameter_value_map["resource"]["value"]
+        #resource = self.resolutions[resource]
+        l.debug(f"provision_container:")
+        l.debug(f" destination: {destination}")
+        l.debug(f" amount: {value} {units}")
+        l.debug(f" resource: {resource}")
+
+        resource_str = f"[{resource.name}]({resource.types[0]})"
+        destination_str = f"`{destination.source.lookup().value.lookup().value.name}({destination.mask})`"
+        self.markdown_steps += [f"Pipette {value} {units} of {resource_str} into {destination_str}."]
+
+
+        return results
+
+    def plate_coordinates(self, record: paml.ActivityNodeExecution):
+        results = {}
+        call = record.call.lookup()
+        parameter_value_map = call.parameter_value_map()
+
+        source = parameter_value_map["source"]["value"]
+        #container = self.var_to_entity[source]
+        coords = parameter_value_map["coordinates"]["value"]
+
+        self.var_to_entity[parameter_value_map['samples']["value"]] = coords
+        l.debug(f"plate_coordinates:")
+        l.debug(f"  source: {source}")
+        l.debug(f"  coordinates: {coords}")
+
+        return results
+
+    def measure_absorbance(self, record: paml.ActivityNodeExecution):
+        results = {}
```

Comment:
We're going to need to think of a way to make the code for libraries be more modular.
It should be possible to add a new library of primitives without having to modify this file. Instead, this file should be able to reach out and find out if there is a specialization for a library that's being used.

Reply:
Add an issue for this so that it doesn't block the merge?